kratos/solving_strategies/schemes/residual_based_bdf2_displacement_scheme.h

```diff
@@ -0,0 +1,600 @@
+
+    typedef typename Element::DofsVectorType                            DofsVectorType;
+
+    typedef typename ImplicitBaseType::TSystemMatrixType             TSystemMatrixType;
+
+    typedef typename ImplicitBaseType::TSystemVectorType             TSystemVectorType;
+
+    typedef typename ImplicitBaseType::LocalSystemVectorType     LocalSystemVectorType;
+
+    typedef typename ImplicitBaseType::LocalSystemMatrixType     LocalSystemMatrixType;
+
+    typedef ModelPart::NodesContainerType                               NodesArrayType;
+
+    typedef ModelPart::ElementsContainerType                         ElementsArrayType;
+
+    typedef ModelPart::ConditionsContainerType                     ConditionsArrayType;
+
+    typedef typename BaseType::Pointer                                 BaseTypePointer;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * Constructor.
+     * The DBF2 method
+     */
+    ResidualBasedBDF2DisplacementScheme()
+        :ImplicitBaseType()
+    {
+    }
+
+    /** Copy Constructor.
+     */
+    ResidualBasedBDF2DisplacementScheme(ResidualBasedBDF2DisplacementScheme& rOther)
+        :ImplicitBaseType(rOther)
+        ,mBDF2(rOther.mBDF2)
+    {
+    }
+
+    /**
+     * Clone
+     */
+    BaseTypePointer Clone() override
+    {
+        return BaseTypePointer( new ResidualBasedBDF2DisplacementScheme(*this) );
+    }
+
+    /** Destructor.
+     */
+    ~ResidualBasedBDF2DisplacementScheme
+    () override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Performing the update of the solution
+     * @details Incremental update within newton iteration. It updates the state variables at the end of the time step u_{n+1}^{k+1}= u_{n+1}^{k}+ \Delta u
+     * @param rModelPart The model of the problem to solve
+     * @param rDofSet Set of all primary variables
+     * @param A LHS matrix
+     * @param Dx incremental update of primary variables
+     * @param b RHS Vector
+     */
+
+    void Update(
+        ModelPart& rModelPart,
+        DofsArrayType& rDofSet,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b 
+        ) override
+    {
+        KRATOS_TRY;
+
+        // Update of displacement (by DOF)
+        const int num_dof = static_cast<int>(rDofSet.size());
+
+        #pragma omp parallel for
+        for(int i = 0;  i < num_dof; ++i) {
+            auto it_dof = rDofSet.begin() + i;
+
+            if (it_dof->IsFree())
+                it_dof->GetSolutionStepValue() += TSparseSpace::GetValue(Dx,it_dof->EquationId());
+        }
+
+        // Updating time derivatives (nodally for efficiency)
+        const int num_nodes = static_cast<int>(rModelPart.Nodes().size());
+
+        #pragma omp parallel for
+        for(int i = 0;  i < num_nodes; ++i) {
+            auto it_node = rModelPart.Nodes().begin() + i;
```

Comment:
this is ok, you don't need to change this. 

However just note that you also have the function NodesBegin()